src/data/set/ncard.lean

```diff
@@ -0,0 +1,630 @@
+
+/-
+Copyright (c) 2023 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import data.finite.card
+
+/-!
+# Noncomputable Set Cardinality
+
+We define the cardinality `set.ncard s` of a set `s` as a natural number. This function is
+noncomputable (being defined in terms of `nat.card`) and takes the value `0` if `s` is infinite.
+
+This is intended as an alternative to `finset.card` and `fintype.card`, both of which contain data
+in their definition that can cause awkwardness when using `set.to_finset`. Using `set.ncard` allows
+cardinality computations to avoid `finset`/`fintype` completely, staying in `set` and letting
+finiteness be handled either explicitly or via typeclasses. An ideal use case is when working
+with terms in `set α` where a `finite α` instance is present.
```

Comment:
Isn't it intended as an alternative to `nat.card`?

Reply:
I guess, yes - I guess I was seeing it more as an API/packaging for `nat.card`.